### ISSUE TITLE
hiddenInput renders label

### ISSUE THREAD

**Issue by vuhrmeister** [yiisoft/yii2#2990]
I'm using `$form->field($model, 'somename')->hiddenInput()` and I would expect that it's completely hidden. Instead, the `<label>` is still rendered. 

An hidden input field does not require any label, imho.

**Comment by spikyjt** [comment]
Sorry to bring this up again, but I really don't think this is a satisfactory resolution. You say you don't want to make a special case for hiddenInput(), but you did for fileInput() (see https://github.com/yiisoft/yii2/pull/795).

The expected behaviour is that hiddenInput() should render a hidden field without labels, hints or errors. I can't think of any use case where you would want these rendered for a hidden field.

A simple fix for this is as follows:

```
public function hiddenInput($options = [])
{
    $options = array_merge($this->inputOptions, $options);
    $this->adjustLabelFor($options);
    $this->parts['{input}'] = Html::activeHiddenInput($this->model, $this->attribute, $options);
    $this->parts['{label}'] = $this->parts['{error}'] = $this->parts['{hint}'] = '';

    return $this;
}
```

Any objection to this? If not, I'll make a PR.

**Comment by qiangxue** [comment]
If you don't need label, hint and error, just use `Html::activeHiddenInput()`. That's the purest thing you can use. As said in my earlier comment, this issue has been brought up back and further several times. So we would like to just keep it as is.

**Comment by spikyjt** [comment]
I hear you, but using `Html::activeHiddenInput()` is not really an answer to this problem. `ActiveField::hiddenInput()` is inherently broken and there are several (closed) issues that marked it as such. Either this function is useless or it should behave correctly. My particular use case is that I have a widget that works like the old form builder for Yii 1.1 and calls the appropriate ActiveField method depending on the config. This process expects the ActiveField methods to work correctly. Of course I can subclass ActiveField, which I have for now, but I don't see why the core should not be fixed. As I said, I can't think of a use case for keeping the label and other parts for a hidden field and none has been mentioned in this issue or others like it.

**Comment by qiangxue** [comment]
Why using `Html::activeHiddenInput()` can't solve your problem?

As I said, some people want this method to generate label/error, while some (like you) don't. We have to make a choice ultimately and cannot just go back and forth forever.

**Comment by spikyjt** [comment]
Because I have to make a special use case in my class (which I hope will be part of the core one day for #59) where I really think it should be done in the core class which is broken.

Where are the people who want this method to generate the label/error? No one has said it should in this thread or others that I've seen. I can't imagine any use case for it, as hidden fields are not displayed in the browser, nor has any such use case even been suggested, let alone demonstrated.

**Comment by qiangxue** [comment]
> Because I have to make a special use case in my class...

Hidden field IS special, according to your need. So it's not a big problem to specially handle it. The only difference is whether we should do it in custom code or core code. The problem of doing it in the core code, however, is that you will no easy way to get the behavior WITH label.

>  I can't imagine any use case for it,..

I can imagine a use case which uses a hidden input and some custom HTML element to get user input (such as an image selector). Behind the scene you use javascript to fill the hidden input and make it to do the validation and show error message, as if the custom HTML element is doing the validation.

**Comment by spikyjt** [comment]
OK, so that is a valid, but unusual use case. It should still certainly not be the default behaviour. I agree though, that you should be able to render the other parts if you so choose. The framework should be flexible, but with sensible defaults. In that case, my suggested patch would look like this:

```php
public function hiddenInput($options = [])
{
    $renderAllParts = ArrayHelper::remove($options, 'renderAllParts', false);
    $options = array_merge($this->inputOptions, $options);
    $this->adjustLabelFor($options);
    $this->parts['{input}'] = Html::activeHiddenInput($this->model, $this->attribute, $options);
    if (!$renderAllParts) {
        $this->parts['{label}'] = $this->parts['{error}'] = $this->parts['{hint}'] = '';
    }
    
    return $this;
}
```

So setting `renderAllParts` to `true` in `options` would render the label etc, but the default would be not to. A better name for that custom option could be suggested of course.

I'd be happy to agree that this might go in 2.1 rather than a 2.0.x milestone, as I know you are a fierce defender of BC! I still think the default behaviour should be standard HTML practice though.

**Comment by qiangxue** [comment]
I'm reopening this for 2.1.x discussion.

**Comment by spikyjt** [comment]
Thanks. I'll create a PR and reference it, ready to be used should the discussion approve it.

**Comment by wangchj** [comment]
I do find the current behavior (showing label) slightly counter-intuitive.

How about inside `widgets\ActiveField` class, we set `$this->parts['{label}'] = '';` as default for hidden fields. For people who wants to show labels, they can call `$this->label($label, $options)`, where if `$label` is true or null, we set the label to model attribute label. I think this can be easily implemented.

Maybe the same problem with field error?

**Comment by WadeShuler** [comment]
It looks like this has been an issue for a while now.

Whatever the "use case" it should be part of the form field, because it is a form element just like the rest of them. Excluding the `hiddenInput()` would cause confusion among many of the users. We would be able to use all the form elements, and have to use `ActiveField::hiddenInput()` for the hidden ones. I think it would add an overhead to people learning Yii, and add confusion since this feature has been in here a while. Most "tutorials" one would search, would include it, but it wouldn't be there when they went to use it.

Since adding `->label(false)` works perfect, the easiest and best solution would be to have `hiddenInput` automatically turn off the label. I would assume 99% of the time, nobody will want a label for a hidden input anyways.

**Comment by HanafiAhmat** [comment]
Extending @spikyjt suggestion. This way, we could give more choices to render other elements.
```php  
public function hiddenInput($options = [])
{
    $renderParts = ArrayHelper::remove($options, 'renderParts', false);
    $options = array_merge($this->inputOptions, $options);
    $this->adjustLabelFor($options);
    $this->parts['{input}'] = Html::activeHiddenInput($this->model, $this->attribute, $options);
    if ($renderParts) {
        if (!in_array('label', $renderParts)) {
            $this->parts['{label}'] = '';
        }
        if (!in_array('error', $renderParts)) {
            $this->parts['{error}'] = '';
        }
        if (!in_array('hint', $renderParts)) {
            $this->parts['{hint}'] = '';
        }
    } else {
        $this->parts['{label}'] = $this->parts['{error}'] = $this->parts['{hint}'] = '';
    }

    return $this;
}
```

* For those who wants default hidden input behavior may use this:
```php
$form->field($model, 'attribute')->hiddenInput()
```

* For those who wants other elements to be rendered as well may use this:
```php
$form->field($model, 'attribute')->hiddenInput(['renderParts'=>['label']])
```

**Comment by c006** [comment]
Hi Hanafi.

From what I can see, this is already built in.
The (c006) ActiveForm has chaining as it is.
But I don't see why any other options are need to hide an element.
Explain if you feel I am missing something.

Thanks,
Jon

**Comment by HanafiAhmat** [comment]
hi @c006,

yes correct. we can achieve desired output where its label will not be rendered by writing this:
$form->field($model, 'attribute')->hiddenInput()->label(false);

However, i prefer the default behavior of hiddenInput() to render without label, error and hint since most of the time i don't need them.

Another reason is that i prefer to write:
$form->field($model, 'attribute')->[fieldFunction]
for all the inputs, instead of a mixture of:
$form->field($model, 'attribute')->[fieldFunction]
and:
Html::activeHiddenInput()
within the same scope for the sake of readability.

If i really need to render any of the extra elements for certain use cases which are not frequent, i can define them as option params as stated in my earlier post.

**Comment by WadeShuler** [comment]
Point being, we shouldn't use a "special hack" or anything to get this to work exactly like it should.

All web browsers do not display a hidden field, that is just normal behavior. No label, nothing, it is just hidden and in the source. The default for Yii, should do exactly that.

One would think `$form->field($model, 'attribute')->hiddenInput()` would do exactly the same. There is *no* reason it should show a label for a hidden field. Hidden means exactly that, hidden. We shouldn't have to add `->label(false)`. If your one of the 1% that wants a label for a hidden field, then you should have to add it.

It isn't about us, we all here know to set the label to false, but your just adding an extra thing to the learning curve for newcomers to Yii. Every wrench you throw into the learning process of Yii, turns more people away. In order to be the "Wordpress" of the frameworks, things like this would need to be addressed to make jumping on Yii as easy and fluid as possible.

Alternating between the two form methods is just refusing to fix the problem correctly.

**Comment by ZlatanNikolich** [comment]
fdf d

**Comment by lennartvdd** [comment]
I agree with @WadeShuler it's just not very intuitive to switch to the Html helper.
Why not simply add `$this->label(false)` inside `ActiveField::hiddenInput()`?:

```php
    public function hiddenInput($options = [])
    {
        $this->label(false);
        $options = array_merge($this->inputOptions, $options);
        $this->adjustLabelFor($options);
        $this->parts['{input}'] = Html::activeHiddenInput($this->model, $this->attribute, $options);

        return $this;
    }
```

Then that 1% that does want the label could always do: 
`$form->field($model, 'fieldname')->hiddenInput()->label('the label')` ?

**Comment by condor-bird** [comment]
I subscribe to the issue. It will be logical if the "label" a hidden field is not displayed by default. Nothing must give aware of the hidden field. In this case, the "label" is irrelevant.

**Comment by Joey33** [comment]
Wade's opinion is 100% mine. Leaving generating labels for hidden fields by default makes Yii unintuitive and actually it's the first time I've felt things in Yii go against logic. Default should be 'hidden means totally hidden'.

**Comment by laurabennett** [comment]
As a new Yii user this is a huge PITA!  What a waste of time trying to figure out how to hide parts of a HIDDEN field!  I completely agree with most of the user comments on this thread.  Hide everything by default and if a handful of people need to show the label then let them be the ones to add label(show) or whatever syntax that makes the most sense.

**Comment by SilverFire** [comment]
Resolved by c64cfa9373867ec72638de4e3e6b97ef4036b481 for branch 2.1

**Comment by davidnewcomb** [comment]
Is this going to get merged into 2.0.x? Please, please, please, with sugar and strawberries on top please. It's a right pain having to look up how to do it each time ;)

**Comment by spikyjt** [comment]
It can't go into 2.0 because it breaks BC.

**Comment by c006** [comment]
Hi JT.

I will take a look at it, have not visited the code in a while.

Jon

**Comment by c006** [comment]
v.2.0.0 has been released.

Regards

**Comment by spikyjt** [comment]
I'm well aware of that! I was replying to the previous poster who was
asking for this to be merged into the 2.0.x release cycle, rather than 2.1
as scheduled. However, it can only go into 2.1, not 2.0.x as it breaks BC.

**Comment by c006** [comment]
Apologies, I thought this was a different thread...
